core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/ServiceProviderBuildItem.java

```diff
@@ -18,6 +27,50 @@
     private final String serviceInterface;
     private final List<String> providers;
 
+    /**
+     * Creates and returns a {@link ServiceProviderBuildItem} for the {@code serviceInterfaceClassName} by including
+     * all the providers that are listed in the service interface descriptor file. The passed classloader is used to
+     * find the service interface descriptor file and if no such file is found, this method returns an
+     * {@link Optional#empty() empty value}.
+     *
+     * @param serviceInterfaceClassName the interface whose service interface descriptor file we want to embed
+     * @param cl the classloader to use to find the service interface descriptor file. Can be null, in which case,
+     *        this method uses the {@link Thread#getContextClassLoader() thread context classloader}
+     * @return
+     * @throws IOException
+     */
+    public static Optional<ServiceProviderBuildItem> allProviders(final String serviceInterfaceClassName, final ClassLoader cl)
+            throws IOException {
+        if (serviceInterfaceClassName == null || serviceInterfaceClassName.trim().isEmpty()) {
+            throw new IllegalArgumentException("service interface name cannot be null or blank");
+        }
+        final ClassLoader classLoaderToUse = cl == null ? Thread.currentThread().getContextClassLoader() : cl;
```

Comment:
An issue with this approach is that if there are additional service implementations on the deployment class path that are not on the runtime class path this will cause failures. Not sure how likely this in practice.

Reply:
You are right. I hadn't thought of this scenario. I think I might have a slightly different way to achieve this same functionality without having to run into the issue that you note here. I'll try it out and update the PR shortly.